routing/grandcentral/server.go

```diff
@@ -138,3 +100,92 @@ func (s *Server) handleMessage(
 
 var _ proxy.RequestHandler = &Server{}
 var _ proxy.Proxy = &Server{}
+
+func getRoutingRecord(ds datastore.Datastore, k util.Key) (*dhtpb.Record, error) {
+	dskey := k.DsKey()
+	val, err := ds.Get(dskey)
+	if err != nil {
+		return nil, errors.Wrap(err)
+	}
+	recordBytes, ok := val.([]byte)
+	if !ok {
+		return nil, fmt.Errorf("datastore had non byte-slice value for %v", dskey)
+	}
+	var record dhtpb.Record
+	if err := proto.Unmarshal(recordBytes, &record); err != nil {
+		return nil, errors.New("failed to unmarshal dht record from datastore")
+	}
+	return &record, nil
+}
+
+func putRoutingRecord(ds datastore.Datastore, k util.Key, value *dhtpb.Record) error {
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return err
+	}
+	dskey := k.DsKey()
+	// TODO namespace
+	if err := ds.Put(dskey, data); err != nil {
+		return err
+	}
+	return nil
+}
+
+func putRoutingProviders(ds datastore.Datastore, k util.Key, newRecords []*dhtpb.Message_Peer) error {
+	log.Event(context.Background(), "putRoutingProviders", &k)
+	oldRecords, err := getRoutingProviders(ds, k)
+	if err != nil {
+		return err
+	}
+	mergedRecords := make(map[string]*dhtpb.Message_Peer)
+	for _, provider := range oldRecords {
+		mergedRecords[provider.GetId()] = provider // add original records
+	}
+	for _, provider := range newRecords {
+		mergedRecords[provider.GetId()] = provider // overwrite old record if new exists
+	}
```

Comment:
important to make it clear in the implementation design doc (maybe in the package's godoc or above this method) that currently there is a last-writer-wins race between GCR nodes.

This should also timeout/expire records. We must not store provider records indefinitely. the current expiration is something like 12/24hrs.